File: backend/src/api/projects.ts
import { Router, Request, Response } from "express";
import multer from "multer";
import { randomUUID } from "crypto";
import { storage } from "../services/StorageService.js";
import { llmService } from "../services/LLMService.js";
import { ragService } from "../services/RAGService.js";
import { logger } from "../logger.js";
import {
  Project,
  ProjectDocument,
  ConversationMessage,
} from "../models/Project.js";

export const router = Router();
const log = logger.child("ProjectsAPI");

// Configure multer for file uploads (memory storage)
const upload = multer({ storage: multer.memoryStorage() });

/**
 * 1. POST /projects
 * Create a new project
 */
router.post("/projects", async (req: Request, res: Response) => {
  try {
    const { name } = req.body as {
      name?: string;
    };

    if (!name || typeof name !== "string" || name.trim() === "") {
      res.status(400).json({ error: "Project name is required" });
      return;
    }

    const project: Project = {
      id: randomUUID(),
      name: name.trim(),
      createdAt: new Date().toISOString(),
    };

    await storage.createProject(project);
    log.info("Project created", { projectId: project.id, name: project.name });

    res.status(201).json({ project });
  } catch (error) {
    log.error("Error creating project", error);
    res.status(500).json({ error: "Failed to create project" });
  }
});

/**
 * GET /projects
 * List all projects
 */
router.get("/projects", async (_req: Request, res: Response) => {
  try {
    const projects = await storage.getAllProjects();
    log.info("Listing projects", { count: projects.length });
    res.json({ projects });
  } catch (error) {
    log.error("Error fetching projects", error);
    res.status(500).json({ error: "Failed to fetch projects" });
  }
});

/**
 * PUT /projects/:id/requirements
 * Update text requirements for a project
 */
router.put(
  "/projects/:id/requirements",
  async (req: Request, res: Response) => {
    try {
      const { id } = req.params;
      const { textRequirements } = req.body as { textRequirements?: string };

      const project = await storage.getProject(id);
      if (!project) {
        res.status(404).json({ error: "Project not found" });
        return;
      }

      project.textRequirements =
        textRequirements && typeof textRequirements === "string"
          ? textRequirements.trim()
          : undefined;
      await storage.updateProjectRequirements(
        id,
        project.textRequirements || ""
      );

      log.info("Updated project requirements", {
        projectId: project.id,
        hasText: Boolean(project.textRequirements),
      });

      res.json({ project });
    } catch (error) {
      log.error("Error updating requirements", error);
      res.status(500).json({ error: "Failed to update requirements" });
    }
  }
);

/**
 * 2. POST /projects/:id/documents
 * Upload documents to a project
 */
router.post(
  "/projects/:id/documents",
  upload.array("files"),
  async (req: Request, res: Response) => {
    try {
      const { id } = req.params;
      const files = req.files as Express.Multer.File[];

      const project = await storage.getProject(id);
      if (!project) {
        res.status(404).json({ error: "Project not found" });
        return;
      }

      if (!files || files.length === 0) {
        res.status(400).json({ error: "No files uploaded" });
        return;
      }

      const documents: ProjectDocument[] = [];

      for (const file of files) {
        // Extract text based on MIME type
        let rawText: string;

        if (file.mimetype.startsWith("text/")) {
          // Plain text files
          rawText = file.buffer.toString("utf-8");
        } else if (file.mimetype === "application/pdf") {
          // Placeholder for PDF extraction
          rawText = `[PDF Document: ${file.originalname}]\n[Text extraction not implemented in POC - placeholder content]`;
        } else if (
          file.mimetype ===
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document" ||
          file.mimetype === "application/msword"
        ) {
          // Placeholder for DOCX/DOC extraction
          rawText = `[Word Document: ${file.originalname}]\n[Text extraction not implemented in POC - placeholder content]`;
        } else {
          rawText = `[File: ${file.originalname}]\n[Unsupported file type for text extraction]`;
        }

        const document: ProjectDocument = {
          id: randomUUID(),
          projectId: id,
          fileName: file.originalname,
          mimeType: file.mimetype,
          rawText,
          uploadedAt: new Date().toISOString(),
        };

        await storage.addDocument(document);
        documents.push(document);
      }

      log.info("Documents uploaded", {
        projectId: id,
        documents: documents.length,
      });

      res.status(201).json({ documents });
    } catch (error) {
      log.error("Error uploading documents", error);
      res.status(500).json({ error: "Failed to upload documents" });
    }
  }
);

/**
 * 3. POST /projects/:id/analyze-docs
 * Analyze all documents and generate initial ProjectState
 */
router.post(
  "/projects/:id/analyze-docs",
  async (req: Request, res: Response) => {
    try {
      const { id } = req.params;

      const project = await storage.getProject(id);
      if (!project) {
        res.status(404).json({ error: "Project not found" });
        return;
      }

      const documents = await storage.getDocuments(id);

      // Collect text sources: documents + text requirements
      const documentTexts = documents.map((doc) => doc.rawText);
      if (project.textRequirements) {
        documentTexts.push(project.textRequirements);
      }

      if (documentTexts.length === 0) {
        res
          .status(400)
          .json({ error: "No documents or text requirements to analyze" });
        return;
      }

      log.info("Document analysis requested", {
        projectId: id,
        documentCount: documentTexts.length,
      });

      // Call LLM to analyze documents
      const projectState = await llmService.analyzeDocuments(documentTexts);
      projectState.projectId = id;

      // Save state
      await storage.saveState(projectState);

      log.info("Document analysis completed", { projectId: id });
      res.json({ projectState });
    } catch (error) {
      log.error("Error analyzing documents", error);
      res.status(500).json({
        error: "Failed to analyze documents",
        details: error instanceof Error ? error.message : String(error),
      });
    }
  }
);

/**
 * 4. POST /projects/:id/chat
 * Send a chat message and get response with updated state
 */
router.post("/projects/:id/chat", async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const { message } = req.body as { message?: string };

    if (!message || typeof message !== "string") {
      res.status(400).json({ error: "Message is required" });
      return;
    }

    const project = await storage.getProject(id);
    if (!project) {
      res.status(404).json({ error: "Project not found" });
      return;
    }

    const currentState = await storage.getState(id);
    if (!currentState) {
      res.status(400).json({
        error: "Project state not initialized. Please analyze documents first.",
      });
      return;
    }

    // Save user message
    const userMessage: ConversationMessage = {
      id: randomUUID(),
      projectId: id,
      role: "user",
      content: message,
      timestamp: new Date().toISOString(),
    };
    await storage.addMessage(userMessage);

    // Get recent conversation history (last 10 messages)
    const recentMessages = await storage.getMessages(id, 10);

    log.info("Processing chat message", {
      projectId: id,
      messageLength: message.length,
      historyCount: recentMessages.length,
    });

    // Process with LLM
    const response = await llmService.processChatMessage(
      message,
      currentState,
      recentMessages
    );

    // Save assistant message
    const assistantMessage: ConversationMessage = {
      id: randomUUID(),
      projectId: id,
      role: "assistant",
      content:
        response.assistantMessage || "I've updated the architecture sheet.",
      timestamp: new Date().toISOString(),
      wafSources: response.sources,
    };
    await storage.addMessage(assistantMessage);

    // Update project state
    await storage.saveState(response.projectState);

    log.info("Chat response generated", {
      projectId: id,
      hasWafSources: Boolean(response.sources && response.sources.length > 0),
    });

    res.json({
      message: assistantMessage.content,
      projectState: response.projectState,
      wafSources: response.sources || [],
    });
  } catch (error) {
    log.error("Error processing chat", error);
    res.status(500).json({
      error: "Failed to process chat message",
      details: error instanceof Error ? error.message : String(error),
    });
  }
});

/**
 * 5. GET /projects/:id/state
 * Get current project state (Architecture Sheet)
 */
router.get("/projects/:id/state", async (req: Request, res: Response) => {
  try {
    const { id } = req.params;

    const project = await storage.getProject(id);
    if (!project) {
      res.status(404).json({ error: "Project not found" });
      return;
    }

    const state = await storage.getState(id);
    if (!state) {
      res.status(404).json({ error: "Project state not initialized" });
      return;
    }

    log.info("Returning project state", { projectId: id });
    res.json({ projectState: state });
  } catch (error) {
    log.error("Error fetching state", error);
    res.status(500).json({ error: "Failed to fetch project state" });
  }
});

/**
 * 6. GET /projects/:id/architecture/proposal
 * Generate Azure architecture proposal with Server-Sent Events for progress
 */
router.get(
  "/projects/:id/architecture/proposal",
  async (req: Request, res: Response) => {
    try {
      const { id } = req.params;

      const project = await storage.getProject(id);
      if (!project) {
        res.status(404).json({ error: "Project not found" });
        return;
      }

      const state = await storage.getState(id);
      if (!state) {
        res.status(400).json({
          error:
            "Project state not initialized. Please analyze documents first.",
        });
        return;
      }

      // Set up SSE headers
      res.setHeader("Content-Type", "text/event-stream");
      res.setHeader("Cache-Control", "no-cache");
      res.setHeader("Connection", "keep-alive");

      const sendProgress = (stage: string, detail?: string): void => {
        const data = { stage, detail, timestamp: new Date().toISOString() };
        res.write(`data: ${JSON.stringify(data)}\n\n`);
      };

      log.info("Generating architecture proposal with SSE", { projectId: id });

      sendProgress("started", "Initializing proposal generation");

      // Create progress callback
      const onProgress = (stage: string, detail?: string): void => {
        log.info("Progress update", { stage, detail });
        sendProgress(stage, detail);
      };

      // Generate architecture proposal with progress updates
      const proposal = await llmService.generateArchitectureProposal(
        state,
        onProgress
      );

      sendProgress("completed", "Proposal generated successfully");

      // Send final result
      res.write(
        `data: ${JSON.stringify({
          stage: "done",
          proposal,
          timestamp: new Date().toISOString(),
        })}\n\n`
      );

      res.end();
      log.info("Proposal generation completed and streamed to client");
    } catch (error) {
      log.error("Error generating proposal", error);

      // Try to send error via SSE if connection is still open
      try {
        res.write(
          `data: ${JSON.stringify({
            stage: "error",
            error: error instanceof Error ? error.message : String(error),
            timestamp: new Date().toISOString(),
          })}\n\n`
        );
        res.end();
      } catch {
        // If SSE connection failed, send regular error response
        if (!res.headersSent) {
          res.status(500).json({
            error: "Failed to generate architecture proposal",
            details: error instanceof Error ? error.message : String(error),
          });
        }
      }
    }
  }
);

/**
 * GET /projects/:id/messages
 * Get conversation history
 */
router.get("/projects/:id/messages", async (req: Request, res: Response) => {
  try {
    const { id } = req.params;

    const project = await storage.getProject(id);
    if (!project) {
      res.status(404).json({ error: "Project not found" });
      return;
    }

    const messages = await storage.getMessages(id);
    log.info("Returning messages", { projectId: id, count: messages.length });
    res.json({ messages });
  } catch (error) {
    log.error("Error fetching messages", error);
    res.status(500).json({ error: "Failed to fetch messages" });
  }
});

/**
 * GET /kb/list
 * List all available knowledge bases
 */
router.get("/kb/list", async (_req: Request, res: Response) => {
  try {
    const kbs = await ragService.listKnowledgeBases();
    log.info("Listing knowledge bases", { count: kbs.knowledge_bases.length });
    res.json(kbs);
  } catch (error) {
    log.error("Error listing knowledge bases", error);
    res.status(500).json({ error: "Failed to list knowledge bases" });
  }
});

/**
 * GET /kb/health
 * Check health status of all knowledge bases
 */
router.get("/kb/health", async (_req: Request, res: Response) => {
  try {
    const health = await ragService.getKBHealth();
    log.info("KB health check", { kbCount: health.knowledge_bases.length });
    res.json(health);
  } catch (error) {
    log.error("Error checking KB health", error);
    res.status(500).json({ error: "Failed to check KB health" });
  }
});

File: backend/app/routers/projects.py
"""
Projects Router - Main project workflow endpoints.
Migrated from TypeScript backend.
"""

from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, delete
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
import json
import logging
from datetime import datetime
import uuid

from app.database import get_db
from app.models import Project, ProjectDocument, ProjectState, ConversationMessage
from app.llm_service import get_llm_service
from app.kb.multi_query import MultiSourceQueryService, QueryProfile
from app.services import get_multi_query_service

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api", tags=["projects"])


# Pydantic models for requests/responses
class CreateProjectRequest(BaseModel):
    name: str


class UpdateRequirementsRequest(BaseModel):
    textRequirements: str


class ChatMessageRequest(BaseModel):
    message: str


class ProjectResponse(BaseModel):
    project: Dict[str, Any]


class ProjectsListResponse(BaseModel):
    projects: List[Dict[str, Any]]


class DocumentsResponse(BaseModel):
    documents: List[Dict[str, Any]]


class StateResponse(BaseModel):
    projectState: Dict[str, Any]


class MessagesResponse(BaseModel):
    messages: List[Dict[str, Any]]


class ChatResponse(BaseModel):
    message: str
    projectState: Dict[str, Any]
    wafSources: List[Dict[str, Any]] = []


# Endpoints

@router.post("/projects", response_model=ProjectResponse)
async def create_project(
    request: CreateProjectRequest,
    db: AsyncSession = Depends(get_db)
):
    """Create a new project."""
    if not request.name or not request.name.strip():
        raise HTTPException(status_code=400, detail="Project name is required")
    
    project = Project(
        id=str(uuid.uuid4()),
        name=request.name.strip(),
        created_at=datetime.utcnow().isoformat()
    )
    
    db.add(project)
    await db.commit()
    await db.refresh(project)
    
    logger.info(f"Project created: {project.id} - {project.name}")
    return {"project": project.to_dict()}


@router.get("/projects", response_model=ProjectsListResponse)
async def list_projects(db: AsyncSession = Depends(get_db)):
    """List all projects."""
    result = await db.execute(select(Project))
    projects = result.scalars().all()
    
    logger.info(f"Listing {len(projects)} projects")
    return {"projects": [p.to_dict() for p in projects]}


@router.put("/projects/{project_id}/requirements", response_model=ProjectResponse)
async def update_requirements(
    project_id: str,
    request: UpdateRequirementsRequest,
    db: AsyncSession = Depends(get_db)
):
    """Update project text requirements."""
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    project.text_requirements = request.textRequirements.strip() if request.textRequirements else None
    await db.commit()
    await db.refresh(project)
    
    logger.info(f"Updated requirements for project: {project_id}")
    return {"project": project.to_dict()}


@router.post("/projects/{project_id}/documents", response_model=DocumentsResponse)
async def upload_documents(
    project_id: str,
    files: List[UploadFile] = File(...),
    db: AsyncSession = Depends(get_db)
):
    """Upload documents to a project."""
    # Check project exists
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    if not files:
        raise HTTPException(status_code=400, detail="No files uploaded")
    
    documents = []
    
    for file in files:
        # Read file content
        content = await file.read()
        
        # Extract text based on MIME type
        if file.content_type and file.content_type.startswith("text/"):
            raw_text = content.decode("utf-8")
        elif file.content_type == "application/pdf":
            raw_text = f"[PDF Document: {file.filename}]\n[Text extraction not implemented in POC - placeholder content]"
        elif file.content_type in [
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            "application/msword"
        ]:
            raw_text = f"[Word Document: {file.filename}]\n[Text extraction not implemented in POC - placeholder content]"
        else:
            raw_text = f"[File: {file.filename}]\n[Unsupported file type for text extraction]"
        
        document = ProjectDocument(
            id=str(uuid.uuid4()),
            project_id=project_id,
            file_name=file.filename or "unknown",
            mime_type=file.content_type or "application/octet-stream",
            raw_text=raw_text,
            uploaded_at=datetime.utcnow().isoformat()
        )
        
        db.add(document)
        documents.append(document)
    
    await db.commit()
    
    logger.info(f"Uploaded {len(documents)} documents to project: {project_id}")
    return {"documents": [d.to_dict() for d in documents]}


@router.post("/projects/{project_id}/analyze-docs", response_model=StateResponse)
async def analyze_documents(
    project_id: str,
    db: AsyncSession = Depends(get_db)
):
    """Analyze documents and generate initial ProjectState."""
    # Get project
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Get documents
    result = await db.execute(
        select(ProjectDocument).where(ProjectDocument.project_id == project_id)
    )
    documents = result.scalars().all()
    
    # Collect text sources
    document_texts = [doc.raw_text for doc in documents]
    if project.text_requirements:
        document_texts.append(project.text_requirements)
    
    if not document_texts:
        raise HTTPException(
            status_code=400,
            detail="No documents or text requirements to analyze"
        )
    
    logger.info(f"Analyzing {len(document_texts)} documents for project: {project_id}")
    
    # Analyze with LLM
    llm_service = get_llm_service()
    state_data = await llm_service.analyze_documents(document_texts)
    
    # Save state
    state_json = json.dumps(state_data)
    
    # Check if state exists
    result = await db.execute(
        select(ProjectState).where(ProjectState.project_id == project_id)
    )
    existing_state = result.scalar_one_or_none()
    
    if existing_state:
        existing_state.state = state_json
        existing_state.updated_at = datetime.utcnow().isoformat()
    else:
        new_state = ProjectState(
            project_id=project_id,
            state=state_json,
            updated_at=datetime.utcnow().isoformat()
        )
        db.add(new_state)
    
    await db.commit()
    
    # Return state with metadata
    state_data["projectId"] = project_id
    state_data["lastUpdated"] = datetime.utcnow().isoformat()
    
    logger.info(f"Document analysis completed for project: {project_id}")
    return {"projectState": state_data}


@router.post("/projects/{project_id}/chat", response_model=ChatResponse)
async def chat_message(
    project_id: str,
    request: ChatMessageRequest,
    db: AsyncSession = Depends(get_db)
):
    """Send a chat message and get response with updated state."""
    if not request.message or not request.message.strip():
        raise HTTPException(status_code=400, detail="Message is required")
    
    # Get project
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Get current state
    result = await db.execute(
        select(ProjectState).where(ProjectState.project_id == project_id)
    )
    state_record = result.scalar_one_or_none()
    
    if not state_record:
        raise HTTPException(
            status_code=400,
            detail="Project state not initialized. Please analyze documents first."
        )
    
    current_state = json.loads(state_record.state)
    
    # Save user message
    user_message = ConversationMessage(
        id=str(uuid.uuid4()),
        project_id=project_id,
        role="user",
        content=request.message,
        timestamp=datetime.utcnow().isoformat()
    )
    db.add(user_message)
    await db.commit()
    
    # Get recent messages
    result = await db.execute(
        select(ConversationMessage)
        .where(ConversationMessage.project_id == project_id)
        .order_by(ConversationMessage.timestamp.desc())
        .limit(10)
    )
    recent_messages = list(reversed(result.scalars().all()))
    recent_message_dicts = [msg.to_dict() for msg in recent_messages]
    
    logger.info(f"Processing chat for project: {project_id}")
    
    # Check if architecture-related (simple keyword detection)
    is_architecture_related = any(
        keyword in request.message.lower()
        for keyword in ["azure", "architecture", "service", "security", "availability", "performance"]
    )
    
    # Query KB if architecture-related
    kb_sources = []
    if is_architecture_related:
        logger.info("Architecture question detected, querying KB")
        try:
            multi_query_service = get_multi_query_service()
            kb_result = multi_query_service.query_profile(
                question=request.message,
                profile=QueryProfile.CHAT,
                top_k_per_kb=3
            )
            if kb_result.get('has_results'):
                kb_sources = kb_result.get('sources', [])
        except Exception as e:
            logger.error(f"KB query failed: {e}")
    
    # Process with LLM
    llm_service = get_llm_service()
    response = await llm_service.process_chat_message(
        request.message,
        current_state,
        recent_message_dicts,
        kb_sources if kb_sources else None
    )
    
    # Save assistant message
    assistant_message = ConversationMessage(
        id=str(uuid.uuid4()),
        project_id=project_id,
        role="assistant",
        content=response['assistantMessage'],
        timestamp=datetime.utcnow().isoformat(),
        waf_sources=json.dumps(response.get('sources', [])) if response.get('sources') else None
    )
    db.add(assistant_message)
    
    # Update project state
    state_record.state = json.dumps(response['projectState'])
    state_record.updated_at = datetime.utcnow().isoformat()
    
    await db.commit()
    
    logger.info(f"Chat response generated for project: {project_id}")
    
    return {
        "message": response['assistantMessage'],
        "projectState": response['projectState'],
        "wafSources": response.get('sources', [])
    }


@router.get("/projects/{project_id}/state", response_model=StateResponse)
async def get_project_state(
    project_id: str,
    db: AsyncSession = Depends(get_db)
):
    """Get current project state."""
    # Check project exists
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Get state
    result = await db.execute(
        select(ProjectState).where(ProjectState.project_id == project_id)
    )
    state_record = result.scalar_one_or_none()
    
    if not state_record:
        raise HTTPException(status_code=404, detail="Project state not initialized")
    
    logger.info(f"Returning state for project: {project_id}")
    return {"projectState": state_record.to_dict()}


@router.get("/projects/{project_id}/architecture/proposal")
async def generate_proposal(
    project_id: str,
    db: AsyncSession = Depends(get_db)
):
    """Generate architecture proposal with Server-Sent Events for progress."""
    # Get project
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Get state
    result = await db.execute(
        select(ProjectState).where(ProjectState.project_id == project_id)
    )
    state_record = result.scalar_one_or_none()
    
    if not state_record:
        raise HTTPException(
            status_code=400,
            detail="Project state not initialized. Please analyze documents first."
        )
    
    state = json.loads(state_record.state)
    
    async def event_generator():
        """Generate SSE events."""
        def send_progress(stage: str, detail: Optional[str] = None):
            data = {
                "stage": stage,
                "detail": detail,
                "timestamp": datetime.utcnow().isoformat()
            }
            return f"data: {json.dumps(data)}\n\n"
        
        yield send_progress("started", "Initializing proposal generation")
        
        try:
            llm_service = get_llm_service()
            
            # Create progress callback
            progress_events = []
            
            def on_progress(stage: str, detail: Optional[str] = None):
                progress_events.append((stage, detail))
            
            # Generate proposal
            proposal = await llm_service.generate_architecture_proposal(state, on_progress)
            
            # Send accumulated progress
            for stage, detail in progress_events:
                yield send_progress(stage, detail)
            
            yield send_progress("completed", "Proposal generated successfully")
            
            # Send final result
            final_data = {
                "stage": "done",
                "proposal": proposal,
                "timestamp": datetime.utcnow().isoformat()
            }
            yield f"data: {json.dumps(final_data)}\n\n"
            
        except Exception as e:
            logger.error(f"Proposal generation failed: {e}")
            error_data = {
                "stage": "error",
                "error": str(e),
                "timestamp": datetime.utcnow().isoformat()
            }
            yield f"data: {json.dumps(error_data)}\n\n"
    
    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive"
        }
    )


@router.get("/projects/{project_id}/messages", response_model=MessagesResponse)
async def get_messages(
    project_id: str,
    db: AsyncSession = Depends(get_db)
):
    """Get conversation history."""
    # Check project exists
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    # Get messages
    result = await db.execute(
        select(ConversationMessage)
        .where(ConversationMessage.project_id == project_id)
        .order_by(ConversationMessage.timestamp.asc())
    )
    messages = result.scalars().all()
    
    logger.info(f"Returning {len(messages)} messages for project: {project_id}")
    return {"messages": [msg.to_dict() for msg in messages]}
